Added feature that let's you specify how often to fetch

hallzy opened the pull request:
I thought that this would be interesting, and useful for some people since 5 minutes may not be often enough for some people, so I made a modification so that you can easily customize how long to wait before fetching.

with --fetch_t you can specify how many seconds to wait before auto fetching.

using the --fetch option gives you the default of 5 minutes still.

"--fetch_t 45" for example would fetch every 45 seconds.

hallzy committed:
Added feature that let's you specify how often to fetch

with --fetch_t you can specify how many seconds to wait before auto fetching.

using the --fetch option gives you the default of 5 minutes still.

"--fetch_t 45" for example would fetch every 45 seconds.

hallzy committed:
Updated readme with --fetch_t feature.

hallzy commented:
Merged changes so that this could be merged easily

michaeldfallen commented on fetch.sh:
```diff
@@ -10,4 +10,8 @@ dot="$(cd "$(dirname "$([ -L "$0" ] && $READLINK_CMD -f "$0" || echo "$0")")"; p
 source $dot/radar-base.sh
 
-fetch;
+if [[ -z "$1" ]]; then
```
Rather than checking $1 you could use $@ (which contains all arguments passed to this script).

hallzy replied:
Agreed. I will change to $@

michaeldfallen commented on radar-base.sh:
```diff
@@ -168,7 +167,14 @@ time_to_update() {
 
 fetch() {
-  if time_to_update; then
+  if [ -z "$1" ]; then
```
Might want to look at default values instead of using an if. It's a way of providing a fallback, e.g.:

```
local foo = ${1:-"some default"}
```

The `:-` separates the variable name and the default to use if the variable isn't defined or is "" (empty string (since all variables in bash are really just strings (except arrays (this is getting silly now)))).

hallzy replied:
Interesting. I did not know that you could do this. I will add this.

michaeldfallen commented on git-radar:
```diff
@@ -83,18 +84,30 @@ if [[ -z $@ ]]; then
+  # Custom fetch of whatever time interval the user wants
+  if [[ "$command" == "--fetch_t" ]]; then
+    # Now $1 is the value of fetch_t
+    nohup $dot/fetch.sh $1 >/dev/null 2>&1 &
```
If you're going to use `$1` here don't forget to shift it, so that it doesn't come back around next iteration. You might want to check it conforms to a number too, in case someone writes `--fetch_t --foo` and you pass `--foo` to your script.

hallzy replied:
Good point. I will do the shift and add a check for a number.

michaeldfallen commented:
Nice change, something I've been meaning to get in there for a while. I've left a few review comments that I think might be worth addressing. I might not be right on those, just starting a conversation.

If doing this I'd also like to follow the pattern the Colour and Format features have followed where they use [`prepare_bash_colors`](https://github.com/michaeldfallen/git-radar/blob/master/radar-base.sh#L16) to set the default time to update, that way you can also use ENV VARS to configure in your `~/.bashrc`.

hallzy commented:
Thanks, I will look into these.

hallzy committed:
Made changes recommeneded by @michaeldfallen

in fetch.sh "if" now checks for $@ instead of just $1

in radar-base.sh there is now a parameter expansion on line 170 instead of the
if statements.

in git-radar, now there is a shift, and a regex check that the next value is a
number. If the next value after --fetch_t is not a number, an error is echo'd
and it resorts to the default behaviour of 5 minutes.

hallzy commented:
I have made the changes recommended... I have not looked into following the pattern that the colour and format features have followed yet since I have a class in 15 min.

hallzy commented:
Also, let me know your input on this, but I was thinking why bother even having --fetch_t ?? Maybe just use --fetch and check for a numeric argument following --fetch... if there is one, then use that... if not then use the default and then we wont need the error message that I put in...

I would like to hear your thoughts...

hallzy commented:
I am not quite sure I understand what you mean by following the pattern of the colour and format features where they used prepare_bash_colors.

Are you thinking of having something like:

```bash
export FETCH_TIME='30'
```

Which we would be able to have in an rc file?

So instead of using the --fetch_t option, using something similar to the above command?

hallzy committed:
Fixed an error in my parameter subsitution.

michaeldfallen commented:
Looks good, thanks for addressing those.

On the pattern I mentioned it goes like this:

in `prepare_bash_colours` and `prepare_zsh_colours` create a new var:

```
FETCH_TIME="${GIT_RADAR_FETCH_TIME:-((5 * 60))}"
```

That will set `FETCH_TIME` to a default of 5 mins unless someone has set `GIT_RADAR_FETCH_TIME` in their `.bashrc` or `.zshrc`.

Then in the `fetch` function:

```
local fetch_time="${1:-$FETCH_TIME}"
...
if time_to_update $fetch_time; then
...
```

So if you pass a time to `fetch` like you do if they specify `fetch_t` then it uses that time, otherwise it falls back to the `GIT_RADAR_FETCH_TIME` and if that isn't set it falls back to the default 5 mins.

(Note: none of this code has been executed, likely there are typos everywhere)

hallzy commented:
Ah... that makes sense... I will work on that now.

hallzy commented:
Okay, I have it working now... Just doing a little bit of testing before pushing it... So with this addition, should we keep the --fetch_t or just get rid of it? Since now we can just do --fetch but have a variable in the bashrc or gitradarrc to specify how long to wait?

The commit I will push for this will get rid of fetch_t and just have fetch. Unless I here back before making the change... If you would prefer to keep the fetch_t, I can add it back afterwards.

hallzy committed:
Custom Fetch time now uses a variable

Set the variable GIT_RADAR_FETCH_TIME in a bashrc, zshrc or gitradarrc file to
customize the fetch time.

hallzy committed:
Missed a fetch_t in the git_radar file to remove.

hallzy committed:
Moved comments back to original position.

hallzy commented:
Changes are in.

Before you merge anything, I can go ahead and squash all of these commits into 1 as well if you would prefer? There has been a lot of changes that have been reverted etc... But if you want to keep the history that is cool too. Just let me know.

hallzy committed:
Since a number is no longer taken after --fetch_t we can remove the parameter
substitution that has been removed in this commit.

michaeldfallen committed:
Merge branch 'master' of github.com:michaeldfallen/git-radar into hallzy_#70

michaeldfallen committed:
Fix tests broken by FETCH_TIME now being pushed in

hallzy committed:
Merge pull request #2 from michaeldfallen/hallzy_#70

Hallzy #70

hallzy commented:
Your PR is merged into this PR

michaeldfallen commented:
:+1: Thanks for this, good determination shown here. Merging, :ship: it!
